File: src/main/java/semicontinuity/idea/code/analyzer/graph/viewModel/Factory.java
package semicontinuity.idea.code.analyzer.graph.viewModel;

import java.util.List;

public interface Factory<
        NODE_PAYLOAD,
        COMP,
        IND_COMPS extends COMP,
        NODE extends COMP,
        SPLIT extends COMP,
        LAYER extends COMP
        > {

    IND_COMPS newIndependentComponents(List<? extends COMP> components);

    NODE newNode(NODE_PAYLOAD payload);

    SPLIT newSplit(List<NODE> items, COMP subLayer);

    LAYER newLayer(COMP directDeps, COMP sharedDeps);
}

File: src/main/java/semicontinuity/idea/code/analyzer/graph/viewModel/Factory.kt
package semicontinuity.idea.code.analyzer.graph.viewModel

interface Factory<VERTEX_PAYLOAD, COMP, IND_COMPS : COMP, VERTEX : COMP, SPLIT : COMP, LAYER : COMP> {

    fun independentCompsIfManyOrNullIfEmpty(items: List<COMP?>): COMP? =
        if (items.isEmpty()) {
            null
        } else {
            independentCompsOrFirst(items)
        }

    fun independentCompsOrFirst(items: List<COMP?>) =
        if (items.size == 1) {
            items[0]
        } else {
            newIndependentComponents(items)
        }

    fun newIndependentComponents(components: List<COMP?>): IND_COMPS

    fun newVertex(payload: VERTEX_PAYLOAD): VERTEX

    fun newSplit(items: List<VERTEX>, subLayer: COMP): SPLIT

    fun newLayer(directDeps: COMP?, sharedDeps: COMP?): LAYER
}
